pennylane/math/quantum.py

```diff
@@ -255,10 +421,38 @@ def to_density_matrix(state, wires, check_state=False):
     [[1.+0.j 0.+0.j]
      [0.+0.j 0.+0.j]]
 
+    >>> y = tf.Variable([1, 0, 0, 0], dtype=tf.complex128)
+    >>> to_density_matrix(z, wires=[1])
+    tf.Tensor(
+    [[1.+0.j 0.+0.j]
+     [0.+0.j 0.+0.j]], shape=(2, 2), dtype=complex128)
+
+    >>> z = [[0.5, 0, 0.5, 0], [0, 0, 0, 0], [0.5, 0, 0.5, 0], [0, 0, 0, 0]]
+    >>> to_density_matrix(z, wires=[0])
+    [[0.5+0.j 0.5+0.j]
+     [0.5+0.j 0.5+0.j]]
+
+    >>> to_density_matrix(z, wires=[1])
+    [[1.+0.j 0.+0.j]
+     [0.+0.j 0.+0.j]]
+
+    >>> y = tf.Variable([[1, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0], [1, 0, 0, 0]], dtype=tf.complex128)
+    >>> to_density_matrix(z, wires=[1])
+    tf.Tensor(
+    [[1.+0.j 0.+0.j]
+     [0.+0.j 0.+0.j]], shape=(2, 2), dtype=complex128)
     """
-    # State vector
-    density_matrix = _density_matrix_from_state_vector(state, wires, check_state)
-    # TODO
-    # Density matrix
     # QNode returning ``qml.state``
+    if isinstance(state, qml.QNode):
+        # TODO: implement for QNode returning state
+        raise NotImplementedError  # pragma: no cover
+    # Cast as a complex128 array
+    state = cast(state, dtype="complex128")
+    len_state = state.shape[0]
+    # State vector
+    if state.shape == (len_state,):
+        density_matrix = _density_matrix_from_state_vector(state, wires, check_state)
+        return density_matrix
+
+    density_matrix = _density_matrix_from_matrix(state, wires, check_state)

```

Comment:
Just wondering if it's good to do an additional check for `state` here? It seems strange that the function delegates to `_density_matrix_from_matrix()` if `state` is a 3D tensor (for example), and the user gets `ValueError: Density matrix must be of shape (2**N, 2**N)`.

Reply:
That's a good point, I would like avoid duplicating the code. I might create a function to check the state format and use it in both functions!

Reply:
I think we should trust that the user is giving the right format and that the check is optional. I don't think we can do it properly directly in this function, because we should create a function that check both for state vector and density matrix.